### PR TITLE
Add logging - default or user-defined

### PR DESCRIPTION
**Pull request by gcfbn**
Code might look better (I think `#[cfg(not(feature = "simple_logging"))]` could be simplified). Generally, is it a good way of using cargo features?

### CODE REVIEW

**Comment by pdodev** [comment]
>I think #[cfg(not(feature = "simple_logging"))] could be simplified
Not really - at least I don't know any other option ...

**Comment by pdodev** [comment]
Please have a look at this: https://crates.io/crates/cfg-if

**Review by pdodev — CHANGES_REQUESTED** [review]
- how would you like to use the DefaultLogger struct and its write_message method ?
- if I don't use feature "flexi_logger", I have to implement the AppLogger trait.
  I think it can be supprising to the user of the utils crate (I got an error message ...).
- I think we need something like this (maybe better name is: simple_logging->default_logging):

```rust
  fn main(&self) -> i32 {
      configure_logging();
  }

  fn configure_logging() {
      #[cfg(feature = "simple_logging")] {
        flexi_logger::Logger::try_with_env().unwrap().log_to_stdout().start().unwrap();
        trace!("Default logger initialized");
      }
  }
```

  In case when simple_logging is not defined we will have an empty implementation of the configure_logging mehtod.
  The user still has the option to implement his own version of this method. 

- please implement an example application:
  RUST_LOG=info cargo run --example use_simple_logging
  RUST_LOG=info cargo run --example use_simple_logging --features simple_logging

  e.g:

```rust

    use anyhow::{Error as aError, Result as aResult};
    use cmd_args::CmdArgs;
    use log::info;
    use thiserror::Error;
    use utils::ApplicationRunner;

    /// Main function that is called when the app starts
    ///
    /// Calls run() function and kills process if it returns an error
    fn main() {
        App.main();
    }

    struct App;

    impl ApplicationRunner for App {
        type CmdArgs = CmdArgs;
        type Error = AppError;

        fn run(&self, _cmd_args: CmdArgs) -> Result<(), Self::Error> {
            println!("hello");
            info!("info -> hello");
            Ok(())
        }
    }

    #[derive(Debug, Error)]
    #[error("App error")]
    struct AppError;
```

examples/use_simple_logging/cmd_args:

```rust
    use clap::{crate_version, AppSettings, Clap};

    #[derive(Clap, Debug)]
    #[clap(version = crate_version!(), setting=AppSettings::ColoredHelp)]
    pub struct CmdArgs {}
```

**Comment by gcfbn** [comment]
> * how would you like to use the DefaultLogger struct and its write_message method ?

I forgot to replace line 23: `flexi_logger::Logger::try_with_env().unwrap().log_to_stdout().start().unwrap();`
with something like this: `DefaultLogger::initialize_logger().unwrap();`

My idea was similar to this:
>In case when simple_logging is not defined we will have an empty implementation of the configure_logging mehtod.
>The user still has the option to implement his own version of this method.

`write_message` method was meant to be used instead of `error!`, `warn!` etc. if someone would like to implement his own logger that isn't compatible with these macros.


>  
> ```
> fn configure_logging() {
>     #[cfg(feature = "simple_logging")] {
>       flexi_logger::Logger::try_with_env().unwrap().log_to_stdout().start().unwrap();
>       trace!("Default logger initialized");
>     }
> }
> ```
>

This way is better, because user doesn't get an error if they don't add `simple_logging` feature to cargo.toml

**Comment by pdodev** [comment]
```sh
❯ cargo check
    Checking utils v0.1.0 (/home/pdo/_pawel/Rust/_2/learning_rust/utils)
error[E0432]: unresolved import `flexi_logger`
 --> src/application_runner.rs:3:5
  |
3 | use flexi_logger::FileSpec;
  |     ^^^^^^^^^^^^ use of undeclared crate or module `flexi_logger`

```

File: utils/Cargo.toml [hunk 1]
```diff
@@ -4,6 +4,13 @@ version = "0.1.0"
 edition = "2018"
 
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+[features]
+default = []
+simple_logging = ["flexi_logger"]
+
+[[example]]
+name = "use_simple_logging"
+path = "examples/use_simple_logging/lib.rs"
```
Comment (pdodev) [inline utils/Cargo.toml:13]: you can remove this line

Comment (pdodev) [inline utils/Cargo.toml:13]: you can remove this line

File: utils/examples/use_simple_logging/lib.rs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    fn run(&self, _cmd_args: CmdArgs) -> Result<(), Self::Error> {
+        println!("hello");
+        info!("info -> hello");
+        Ok(())
+    }
+}
+
+#[derive(Debug, Error)]
+#[error("App error")]
+struct AppError;
```
Comment (pdodev) [inline utils/examples/use_simple_logging/lib.rs:27]: rename lib.rs -> main.rs

File: utils/src/lib.rs [hunk 3]
```diff
@@ -1,3 +1,6 @@
+#[macro_use]
+extern crate log;
+
```
Comment (pdodev) [inline utils/src/lib.rs:3]: remove this code and use 'use' statement.
e.g.: add this code:
```rust
use log::{error,info,trace};
```
to application_runner.rs

File: utils/Cargo.toml [hunk 4]
```diff
@@ -4,6 +4,13 @@ version = "0.1.0"
 edition = "2018"
 
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+[features]
+default = []
+simple_logging = ["flexi_logger"]
+
+[[example]]
```
Comment (pdodev) [inline utils/Cargo.toml:11]: add this line to this section:
```toml
required-features = [ "simple_logging" ]
```

File: utils/Cargo.toml (hunk 1 above)
Comment (pdodev) [inline utils/Cargo.toml:13]: you can remove this line

**Pull request merged**